### PR TITLE
Elf/MachO/COFF binary files loader backend.

### PR DESCRIPTION
**Pull request by feseal**
We have made the BAP plugin for loading Elf, MachO and COFF binaries. We are not completely sure of the correctness of the extraction of the parameters for MachO and COFF. It would be great to hear Your comments.

### CODE REVIEW

File: lib/bap_image/llvm_binary_stubs.cpp [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    CAMLlocal1(result);
+    result = caml_alloc(7, 0);  
+    Store_field (result, 0, caml_copy_string(s.name().c_str()));
+    Store_field (result, 1, caml_copy_int64(s.offset()));
+    Store_field (result, 2, caml_copy_int64(s.addr()));
+    Store_field (result, 3, caml_copy_int64(s.size()));
+    Store_field (result, 4, Val_bool(s.is_readable()));
+    Store_field (result, 5, Val_bool(s.is_writable()));
+    Store_field (result, 6, Val_bool(s.is_executable()));
+    return result;
```
Comment (ivg) [inline lib/bap_image/llvm_binary_stubs.cpp]: You should use `CAMLreturn` in allocating stubs

File: lib/bap_image/llvm_binary_stubs.cpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    case AT::nvptx64: return ::caml_hash_variant("nvptx64");
+    default:
+        std::string message("Unknown arch: ");
+        message += std::to_string(static_cast<int>(arch));
+        ::caml_failwith(message.c_str());
+    }
+}
+
+
+::value to_value(const seg::segment& s) {
```
Comment (ivg) [inline lib/bap_image/llvm_binary_stubs.cpp]: All allocating function must start with `CAMLparam*` macro:

   > The function body must start with one of the [CAMLparam] macros.
   If the function has no parameter of type [value], use [CAMLparam0].
   If the function has 1 to 5 [value] parameters, use the corresponding
   [CAMLparam] with the parameters as arguments.
   If the function has more than 5 [value] parameters, use [CAMLparam5]
   for the first 5 parameters, and one or more calls to the [CAMLxparam]
   macros for the others.
   If the function takes an array of [value]s as argument, use
   [CAMLparamN] to declare it (or [CAMLxparamN] if you already have a
   call to [CAMLparam] for some other arguments).

   > If you need local variables of type [value], declare them with one
   or more calls to the [CAMLlocal] macros at the beginning of the
   function, after the call to CAMLparam.  Use [CAMLlocalN] (at the
   beginning of the function) to declare an array of [value]s.

File: lib/bap_image/llvm_binary_stubs.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    };
+    ::value v = ::alloc_custom(&binary_ops, sizeof(Binary*), 0, 1);
+    *reinterpret_cast<Binary**>(Data_custom_val(v)) = b;
+    return v;
+}
+
+
+::value to_value(Triple::ArchType arch) {
+    typedef Triple::ArchType AT;
+    switch(arch) {
```
Comment (ivg) [inline lib/bap_image/llvm_binary_stubs.cpp]: It is better to use strings here. For example, what if some of the variants has clashing hashes?

File: lib/bap_image/llvm_binary_stubs.cpp [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+namespace impl {
+
+using namespace llvm;
+using namespace llvm::object;
+
+Binary* from_value (::value v) {
+    CAMLparam1(v);
+    return *reinterpret_cast<Binary**>(Data_custom_val(v));
```
Comment (ivg) [inline lib/bap_image/llvm_binary_stubs.cpp]: There still should be `CAMLreturn`. `CAMLparam` must always match with `CAMLreturn`

Comment (ivg) [inline lib/bap_image/llvm_binary_stubs.cpp]: But in this function you can just remove `CAMLparam`, since it allocates nothing.

**Comment by dbrumley** [comment]
I didn't see any test cases. Can you please let us know what the status of testing is? I would like to see tested on macho, PE (when you say coff, you mean just coff or pe/coff?), and elf compiled with gcc and llvm if possible. We have various git repos set up with some testing at https://github.com/binaryanalysisplatform/. Thanks!

**Comment by feseal** [comment]
We have simple test, which just read a file and print content of Img.t record. We tried to read macho ls, mv and paint.exe from windows. We will prepare set of required test after discussing it with @ivg.

**Comment by ivg** [comment]
@tiffanyb, do you have a set of test files for mac and windows platforms? Is there is a chance, that this [script](https://github.com/BinaryAnalysisPlatform/x86_64-binaries/blob/master/build.py) will work on other platforms?

**Comment by feseal** [comment]
Closed for testing

**Pull request closed without merging**